[Serve][Doc] Multiplexing doc (#35701)
ray-project/ray: doc/source/serve/model-multiplexing.md

# Model Multiplexing

> This section helps you understand how to write multiplexed deployment by using new introduced API `serve.multiplexed` and `serve.get_multiplexed_model_id`.

Reviewer: ```suggestion
This section helps you understand how to write multiplexed deployment by using the API `serve.multiplexed` and `serve.get_multiplexed_model_id`.
```

Reviewer: Does it make sense to link to the API ref entries for the API calls?

> This is experimental feature and the API may change in the future, welcome to try it out and give us feedback!

Reviewer: ```suggestion
This is an experimental feature and the API may change in the future. You are welcome to try it out and give us feedback!
```

> ## Why do we need model multiplexing?

Reviewer: ```suggestion
## Why model multiplexing?
```

> Model multiplexing is a technique to serve multiple models in a single replica, traffic will be routed to the corresponding model based on the request header.

Reviewer: ```suggestion
Model multiplexing is a technique for serving multiple models in a single replica. Traffic is routed to the corresponding model based on the request header.
```

> Model multiplexing is a technique to serve multiple models in a single replica, traffic will be routed to the corresponding model based on the request header.

Reviewer 2: Can we change this to - "to serve multiple models with a pool of replicas to optimize cost. This is useful in cases where you might have many models with the same shape but different weights that are sparsely invoked. If any replica for the deployment has the model loaded, incoming traffic for that model (based on request header) will automatically be routed to that replica avoiding unnecessary load time."

> In the previous version of Ray Serve, if you want to serve many models, you have to deploy multiple applications or write routing logics inside your own deployment codes. This is not efficient when you have many models with similar input types and you don't need to load them all unless there are requests asking for them.

Reviewer: ```suggestion
In the previous version of Ray Serve, to serve many models, you deploy multiple applications or write routing logic inside your deployment code. This design is not efficient when you have many models with similar input types and you only need to load them all if requests ask for them.
```

> ## How to write a multiplexed deployment?

Reviewer: ```suggestion
## Writing a multiplexed deployment
```

> To write a multiplexed deployment, you need to use the new introduced API `serve.multiplexed` and `serve.get_multiplexed_model_id`.

Reviewer: ```suggestion
To write a multiplexed deployment, use the `serve.multiplexed` and `serve.get_multiplexed_model_id` APIs.
```

> Assume you have multiple torch models inside aws s3:

Reviewer: ```suggestion
Assuming you have multiple Torch models in AWS S3:
```

```
s3://my_bucket/1/model.pt
s3://my_bucket/2/model.pt
s3://my_bucket/3/model.pt
s3://my_bucket/4/model.pt
...
```

> Define a Multiplexed deployment:

Reviewer: ```suggestion
Define a multiplexed deployment:
```

```{literalinclude} doc_code/multiplexed.py
:language: python
:start-after: __serve_deployment_example_begin__
:end-before: __serve_deployment_example_end__
```

:::{note}

> serve.multiplexed API is also parameterized by `max_num_models_per_replica`. You can use them to configure how many models you want to load in a single replica. If the number of models is larger than `max_num_models_per_replica`, Serve will use LRU policy to evict the least recently used model.

Reviewer: ```suggestion
The `serve.multiplexed` API also has a `max_num_models_per_replica` parameter. Use it to configure how many models to load in a single replica. If the number of models is larger than `max_num_models_per_replica`, Serve uses the LRU policy to evict the least recently used model.
```

:::

:::{tips}

> In the code example, we are using the Pytorch Model object, you can also define your own model class and use it here. You must make sure the model class has `__call__` method, and if you want to release resources when the model is evicted, you can implement `__del__` method. Ray Serve internally will call `__call__` method to execute the model, and call `__del__` method to release resources when the model is evicted.

Reviewer: ```suggestion
This code example uses the Pytorch Model object. You can also define your own model class and use it here. Make sure the model class has a `__call__` method. To release resources when the model is evicted, implement the `__del__` method. Ray Serve internally calls the `__call__` method to execute the model, and calls the `__del__` method to release resources when the model is evicted.
```

:::

> In the code, you can use `serve.get_multiplexed_model_id` to get the model id from each request, ray serve will use this model id to route the request to the corresponding replica.

Author: that is a good point. rephrased and added a note to explain the routing.

To send a request to a specific model, you can use the following code:
```{literalinclude} doc_code/multiplexed.py
:language: python
:start-after: __serve_request_send_example_begin__
:end-before: __serve_request_send_example_end__
```

:::{note}
`serve_multiplexed_model_id` is required in the request header, and the value should be the model id you want to send the request to.

> :::

Author: make sense

After you run the above code, you will see the following output inside the deployment log:
```
INFO 2023-05-24 01:19:03,853 default_Model default_Model#EjYmnQ CUpzhwUUNw / default replica.py:442 - Started executing request CUpzhwUUNw
INFO 2023-05-24 01:19:03,854 default_Model default_Model#EjYmnQ CUpzhwUUNw / default multiplex.py:131 - Loading model '1'.
INFO 2023-05-24 01:19:04,859 default_Model default_Model#EjYmnQ CUpzhwUUNw / default replica.py:542 - __CALL__ OK 1005.8ms
```

> If you load multiple different models in a single replica, and exceed the `max_num_models_per_replica`, you will see the following output inside the deployment log, the least recently used model will be evicted:

Reviewer 3: ```suggestion
If you continue to load more models, and exceed the `max_num_models_per_replica`, the least recently used model will be evicted and you will see the following lines in the deployment logs:
```

# Model Multiplexing

This section helps you understand how to write multiplexed deployment by using the `serve.multiplexed` and `serve.get_multiplexed_model_id` APIs.

This is an experimental feature and the API may change in the future. You are welcome to try it out and give us feedback!

## Why model multiplexing?

> Model multiplexing is a technique for serving multiple models in a single replica. Traffic is routed to the corresponding model based on the request header. To serve multiple models with a pool of replicas to optimize cost. This is useful in cases where you might have many models with the same shape but different weights that are sparsely invoked. If any replica for the deployment has the model loaded, incoming traffic for that model (based on request header) will automatically be routed to that replica avoiding unnecessary load time.

Reviewer 2: Model multiplexing is a technique for serving multiple models in a single replica. -> This is not entirely accurate right? It's to serve multiple similar models with a pool of replicas.

Author: rephrased.

## Wrting a multiplexed deployment

To write a multiplexed deployment, use the `serve.multiplexed` and `serve.get_multiplexed_model_id` APIs.

> Assuming you have multiple Torch models inside aws s3 bucket with the following structure:

Reviewer 3: ```suggestion
Assuming you have multiple Torch models inside an aws s3 bucket with the following structure:
```

```
s3://my_bucket/1/model.pt
s3://my_bucket/2/model.pt
s3://my_bucket/3/model.pt
s3://my_bucket/4/model.pt
...
```

Define a multiplexed deployment:
```{literalinclude} doc_code/multiplexed.py
:language: python
:start-after: __serve_deployment_example_begin__
:end-before: __serve_deployment_example_end__
```

:::{note}
The `serve.multiplexed` API also has a `max_num_models_per_replica` parameter. Use it to configure how many models to load in a single replica. If the number of models is larger than `max_num_models_per_replica`, Serve uses the LRU policy to evict the least recently used model.
:::

:::{tip}

> This code example uses the Pytorch Model object. You can also define your own model class and use it here. Make sure the model class has a `__call__` method. To release resources when the model is evicted, implement the `__del__` method. Ray Serve internally calls the `__call__` method to execute the model, and calls the `__del__` method to release resources when the model is evicted.

Reviewer 2: What is the __call__ method requirement in the model class?

Author: good catch

:::

> `serve.get_multiplexed_model_id` is to retrieve the model id from the request header, and the model_id is supposed to be passed into the `get_model` function. If the model id is not found in the replica, Serve will load the model from the s3 bucket and cache it in the replica. If the model id is found in the replica, Serve will return the cached model.

Reviewer 3: ```suggestion
`serve.get_multiplexed_model_id` is used to retrieve the model id from the request header, and the model_id is then passed into the `get_model` function. If the model id is not found in the replica, Serve will load the model from the s3 bucket and cache it in the replica. If the model id is found in the replica, Serve will return the cached model.
```

> `serve.get_multiplexed_model_id` is to retrieve the model id from the request header, and the model_id is supposed to be passed into the `get_model` function. If the model id is not found in the replica, Serve will load the model from the s3 bucket and cache it in the replica. If the model id is found in the replica, Serve will return the cached model.

Reviewer 2: Can we also describe how the eviction actually happens?

:::{note}

> Internally, serve router will route the traffic to the corresponding replica based on the model id in the request header.

Reviewer 2: Suggest adding - If all replicas holding the model are over-subscribed, it sends the request to a new replica that doesn't have the model loaded (preferring replicas which have not reached capacity in terms of number of models loaded).

Author: added! not add `preferring` sentence, currently we go for round robin there is no special pick up preference right now. We can add it if we see this requirements in the future.
